### PR TITLE
Distance finding

### PR DESCRIPTION
**Pull request by oshlern**
This code finds distance and angle to goal and publishes it to network tables. It is necessary for running with robot code. It has been tested and works. While it may still be inaccurate sometimes, it's definitely better than master. Additionally, it has the unused features of calculating the exact x and y displacements of the robot as well as it's absolute direction. New-distance is integrated with our 3 mains, and Printing has been modified to make readability with distance easier.

### CODE REVIEW

File: PiMain.py [hunk 1]
```diff
@@ -26,15 +26,15 @@ def main():
 	while True:
 		image = PiCamera.getImage()
 		contours = GripRunner.run(image)
-		targets = filterContours(contours) # To be edited if the last filter is changed in case of algorithmic changes. 
-		center = findCenter(targets) #if 2, join and find center, if 1, return val, if 0 return input. if adjustCoords:	center[0] -= halfWidth
+		targets = filterContoursFancy(contours)
+		isVisible, angleToGoal, distance = findSpike(targets)
```
Comment (c0d3rman) [inline PiMain.py]: Remove trailing whitespace

File: SpikeFinding.py [hunk 2]
```diff
@@ -1,9 +1,72 @@
...
+def distanceFromHeight(y):
+	degrees = np.multiply(config.degPerPxl, np.subtract(np.true_divide(config.resolution[1], 2), y))
+	degrees = np.add(degrees, config.cameraTilt)
+	distance = np.divide(config.displacement, np.tan(degrees))
+	return distance
+
+def trueDistance(d1, d2):
+	squares = np.add(np.square(d1), np.square(d2))
+	squared = np.subtract(np.multiply(2, squares), np.square(config.width))
+	# print squared, "not negative?" #SHOULDN"T BE NEGATIVE IF CAMERA MOUNTED PROPERLY AND CONSTANTS CORRECTLY LABELED
```
Comment (c0d3rman) [inline SpikeFinding.py]: Remove

File: SpikeFinding.py [hunk 3]
```diff
@@ -1,9 +1,72 @@
...
+	# d = 1/2 * sqrt(2*(d1^2+d2^2)-w^2)
+
+def angle(d, d2):
+	squares = np.subtract(np.add(np.true_divide(np.square(config.width), 4), np.square(d)), np.square(d2))
+	phi = np.arccos(np.divide(squares, np.multiply(config.width, d)))
+	return np.subtract(np.pi, phi)
+	# angle = pi - acos(1/4*w^2 + d^2 - d2^2 / wd)
+
+
+# def widthFromData(d1, d2, pxlsBetween):
```
Comment (c0d3rman) [inline SpikeFinding.py]: Remove all commented out unused code

File: WebCam.py [hunk 4]
```diff
@@ -22,3 +22,8 @@ def set(exposure=False, gain=False, contrast=False):
 
 def getExposure():
 	return int(subprocess.check_output("/usr/bin/v4l2-ctl -d /dev/video0 -C exposure_absolute", shell=True)[19:].strip())
+
+def getResolution():
+	resolution = getImage().shape
+	print resolution
```
Comment (c0d3rman) [inline WebCam.py]: Why is this necessary in `master`?

File: WebCamMain.py [hunk 5]
```diff
@@ -9,41 +9,43 @@
...
+				Printing.display(image, defaultSize=True)
 		try:
-			NetworkTabling.publishToTables(center, frameNum=frameNum)
+			NetworkTabling.publishToTables(isVisible=isVisible, angleToGoal=angleToGoal, distance=distance, frameNum=frameNum)
 		except Exception as error:
-			if debug:
+			if config.debug:
 				print error
-				print "The networktables are mean to us"
+				# print "The networktables are mean to us"
```
Comment (c0d3rman) [inline WebCamMain.py]: Remove

File: WebCamMain.py [hunk 6]
```diff
@@ -9,41 +9,43 @@
...
 		frameNum += 1
+		#cv2.waitKey(500)
```
Comment (c0d3rman) [inline WebCamMain.py]: Remove

**Review by c0d3rman — DISMISSED** [review]

**Comment by oshlern** [comment]
Tested degrees to goal, and it was accurate

File: SpikeFinding.py [hunk 7]
```diff
@@ -1,9 +1,55 @@
...
+import config
 
-def findCenter(contours):
-	if len(contours) == 0:
-		return False
+def findSpike(contours): # returns isVisible, angleToGoal, distance
+	isVisible=False
+	numContours = len(contours)
+	if numContours == 0:
+		print "no contours"
```
Comment (CaseyManning) [inline SpikeFinding.py]: Delete or move to Printing.py?

File: SpikeFinding.py [hunk 8]
```diff
@@ -1,9 +1,55 @@
...
+		return False, 0, 0
 	contour = np.concatenate(contours)
-	x,y,w,h = cv2.boundingRect(contour)
-	return (x + w/2, y + h/2)
+	X,Y,W,H = cv2.boundingRect(contour)
+	center = (np.add(X, np.divide(W,2)), np.add(Y, np.true_divide(H,2)))
+	print np.degrees(np.multiply(config.degPerPxl[0], np.subtract(np.true_divide(config.resolution[0], 2), 0)))
```
Comment (CaseyManning) [inline SpikeFinding.py:13]: Osher wants to put this in Printing.py instead.

File: SpikeFinding.py [hunk 9]
```diff
@@ -1,9 +1,55 @@
...
+	angleToGoal = np.multiply(config.degPerPxl[0], np.subtract(np.true_divide(config.resolution[0], 2), center[0]))
+	if numContours == 2:
+		isVisible = True
+		x1,y1,w1,h1 = cv2.boundingRect(contours[0])
```
Comment (CaseyManning) [inline SpikeFinding.py]: minarearect?

Comment (CaseyManning) [reply SpikeFinding.py]: -Osher

**Review by CaseyManning — CHANGES_REQUESTED** [review]
Seems Good

**Review by CaseyManning — APPROVED** [review]

**Pull request merged**